### ISSUE TITLE
Signed transaction data object

### ISSUE THREAD

**Issue by senexi** [openid/OpenID4VP#342]
Currently the `transaction_data` object is defined as a "base64url encoded JSON object". For some use-cases it might be useful to actually sign this data, so it can later be linked to its originator. So the idea would be to also allow an encoded JWT here.

**Comment by jogu** [comment]
Hi @senexi , can you confirm if you've signed an OpenID contribution agreement please, and if not sign one please? https://openid.net/intellectual-property/openid-foundation-contribution-agreements/

Thanks!

**Comment by c2bo** [comment]
> Currently the `transaction_data` object is defined as a "base64url encoded JSON object". For some use-cases it might be useful to actually sign this data, so it can later be linked to its originator. So the idea would be to also allow an encoded JWT here.

Can you describe your requirement / use-case in a bit more detail? We already have the option to sign the Authorization Request containing the `transaction_data` (which will very likely be required in most use-cases that use transaction_data) - wouldn't that be enough or do you want a second signature by another party?

**Comment by senexi** [comment]
@jogu we are checking the paperwork today and sign it.

**Comment by senexi** [comment]
@c2bo we are currently working on using this feature for payments where the presentation of a credential is requested by a merchant but will be forwarded to an issuing bank for verification. If the `transaction_data` object would be signed, the issuing bank could verify who presented the data to the holder, given the presentation and the original (signed) `transaction_data_object`. The Authorization Request itself does not help in this case.

**Comment by senexi** [comment]
> Hi @senexi , can you confirm if you've signed an OpenID contribution agreement please, and if not sign one please? https://openid.net/intellectual-property/openid-foundation-contribution-agreements/
> 
> Thanks!

@jogu I signed the agreement.

**Comment by senexi** [comment]
> Hi @senexi , can you confirm if you've signed an OpenID contribution agreement please, and if not sign one please? https://openid.net/intellectual-property/openid-foundation-contribution-agreements/
> 
> Thanks!

I signed the paper today.

**Comment by senexi** [comment]
@jogu  I signed the paper today.

> Hi @senexi , can you confirm if you've signed an OpenID contribution agreement please, and if not sign one please? https://openid.net/intellectual-property/openid-foundation-contribution-agreements/
> 
> Thanks!

**Comment by jogu** [comment]
Thanks @senexi !

Can you expand on this bit please:

>  the issuing bank could verify who presented the data to the holder

A flow diagram might help as I'm not sure how the issuing bank and the merchant relate to the issuer/wallet/verifier roles in OID4VP.

**Comment by senexi** [comment]
> Thanks @senexi !
> 
> Can you expand on this bit please:
> A flow diagram might help as I'm not sure how the issuing bank and the merchant relate to the issuer/wallet/verifier roles in OID4VP.

Hi @jogu 

[Here](https://github.com/nobid-consortium/payment-reference-documentation/blob/main/payment-reference-doc.md) it is explained in more detail. 

However the important part might be this. An sd-jwt presentation if forwarded from the verifier to the original issuer of the credential. And the issuer could verify who the originator of the transaction_data object actually is, if it would be signed.  

![image](https://github.com/user-attachments/assets/282dd1ad-abbd-42b5-bc3c-7c49b17d8073)

**Comment by tlodderstedt** [comment]
The current flow design is intended to be used by the ASPSP to authorize a transaction with its customer (through the wallet). That's why a payment credential (issued by the ASPSP) is used to do the authorization. 

Your flow is the typical Open Banking Payment Initiation flow, where the TPP gets an access token to initiate a payment with the ASPSP. I don't understand what the value/purpose of the TPP to wallet interaction is in this case. Can you please explain?

Tp me it seems you want to utilize the credential presentation (with the transaction data) as kind of an access token towards the ASPSP.

**Comment by senexi** [comment]
@stefan-kauhaus yes, looking at open banking, the TPP is identified by the according certificates. However we are not only looking at open banking. If we look at it from a more generic point of view, we are talking about forwarding the presentation - may be even relaying it through multiple parties. So it might not always be the case, that the party submitting the presentation to the issuing ASPSP is the one who created the authorization request for the wallet. In such a case, a signed `transaction_data` object would we a way to maintain this link.

**Comment by cyberphone** [comment]
It is a bit hard to understand what the TPP adds to the pudding in the depicted scenario.  If there is a Merchant involved, it needs to be included in the sequence diagram as well.

In case a TPP forwards an authorization (like illustrated), I would assume that it would _embed_ the received authorization data, in a specific TPP-signed TPP->ASPSP message.

**Comment by cyberphone** [comment]
This is probably the motivation for signed `transaction_data`:
https://www.linkedin.com/posts/andersrundgren_since-payments-are-not-particularly-well-activity-7270734001852620800-gI1K/

**Comment by tlodderstedt** [comment]
Hi all, 

I think the current proposal has some drawbacks. 
- the TPP does need to know claims must be disclosed in the VP
- the VP is not protected against replay between TPP and ASPSP
- the TPP get's to know the content of the VC/VC even though

I suggest to adopt a more Open Banking style approach and use the wallet to issue a sender constrained access token (which contains the payment VP). 

Here is the sequence diagram: 
![image](https://github.com/user-attachments/assets/1332d3c7-2ab4-4d55-8b67-3d32304b7f58)

**Comment by cyberphone** [comment]
Thanx @tlodderstedt, this is AFAIK the first (ever) attempt by somebody representing OpenID, taking a stab at a wallet associated with Open Banking!

Rather than going into the specification details, I believe we should first spend time figuring out what our expectations may be with respect to functionality, security, and privacy.  None of the published pilots have bothered much with these rather fundamental issues.  eIDAS or ARF compliance means little to the payment community.  It is also worth keeping in mind that the competition is fierce and that the market is [rightfully] quite reluctant adopting unproven solutions.   Making a slightly better mousetrap is of course an option, but that doesn't sound very cool.

There is plenty of room for innovation since Apple Pay does _currently_ not support the following: 
✅ Digital receipts. "Digitization" anybody?
✅ Unified POS/Online Payer authorization concept and protocol.
✅ Unified Payer authorization and UX. An account is an account, be it VISA, SEPA, or "XYZ". Let the backend (aka "payment orchestration") deal with that as is already the case for card payments.
✅ Person-to-Person payments using SEPA Inst.

Regarding the specifications themselves, I have a hard time figuring out where the Merchant are.  Merchants are neither relying parties, nor in need of credential presentations; they depend on assurances that they will be (or are already being) paid.  Such data can only be provided by payment backends.  In the rare occasion a Merchant is also a TPP, it shouldn't change anything protocol-wise; it just means taking on multiple roles.

@jogu @Sakurann

**Comment by senexi** [comment]
@tlodderstedt, thanks for picking this up. A few remarks regarding the proposal. As you already mentioned, wrapping the P2Pay in an Access token introduces a new / custom flow for the wallet. Although we also included a proposal for a possible endpoint (A2Pay Direct) where a bank could directly receive a P2Pay, we always envisioned it as something optional as there might be many different / proprietary ways to transport the P2Pay in the end. For example, the signed body feature in the current Berlin Group specs might also be a viable option. Taking this into account, it would mean introducing a specific frontend flow (for all wallets) only to support one possible backend flow, wouldn’t it?

But wouldn’t it be possible to keep the default OpenID4VP flow and tackle the drawbacks you mentioned in a different way.

Couldn’t the ASPSP easily detect a replay by tracking the hash of the P2Pay? Once a P2Pay is processed, the ASPSP would refuse the processing if it would be submitted again. To limit the timespan the ASPSP needs to keep track of the P2Pay hash, the lifetime of the according KB-JWT could be limited by including the `exp` claim into the KB-JWT, making it a rather short-lived token. 

Also sharing the claims of the A2Pay with the TPP might not really be an issue, as the A2Pay should contain only information that actually need to be shared with the TPP.  Like some kind of an account identifier and / or meta data required for processing by the payment rail that the A2Pay is connected with. Also wouldn't the TPP always be able to see those claims.

What we are currently not able to do is, to keep the link between the original TPP (who presented the payment request to the user) and the P2Pay. Although it might not even be required (depending on the payment rail), it would be useful to have the option to keep this link cryptographically - which is why I submitted this feature request here ;-) If the TPP is actually signing the transaction_data object, it would establish this hard link between the wallet signed P2Pay and the TPP, no matter how the P2Pay is relayed to the issuer in the end.

**Comment by senexi** [comment]
@cyberphone 
> None of the published pilots have bothered much with these rather fundamental issues. eIDAS or ARF compliance means little to the payment community. 

Would you maybe care to share some details and explain in which way the LSPs never looked at privacy and security? It is actually a bold statement, considering the fact that the flow proposed by the LSP completely builds on the default OpenID4VP flow for SCA. But maybe there is an issue there…

Regarding the merchant. In this model, the role of the merchant and TPP are the same - as you noticed already. And yes, it should not change anything protocol wise.

**Comment by cyberphone** [comment]
@senexi Pardon me for using somewhat sweeping comments.

> Would you maybe care to share some details and explain in which way the LSPs never looked at privacy and security? It is actually a bold statement, considering the fact that the flow proposed by the LSP completely builds on the default OpenID4VP flow for SCA. But maybe there is an issue there…

Yes, because these specifications do assumptions that doesn't seem to translate well to consumer-to-business payment transactions.

> Regarding the merchant. In this model, the role of the merchant and TPP are the same - as you noticed already. And yes, it should not change anything protocol wise.

Well, this was not my intention.  Merchants and TPPs represent entities having distinct roles as well as unique security, privacy and trust characteristics.  That a single entity can take on multiple roles is another thing.

**Comment by cyberphone** [comment]
@senexi wrote:
> Couldn’t the ASPSP easily detect a replay by tracking the hash of the P2Pay? Once a P2Pay is processed, the ASPSP would refuse the processing if it would be submitted again. To limit the timespan the ASPSP needs to keep track of the P2Pay hash, the lifetime of the according KB-JWT could be limited by including the exp claim into the KB-JWT, making it a rather short-lived token.

Yes!  This is idempotent operation and may even contribute to reliability since it makes retransmission a "thing":
https://fido-web-pay.github.io/specification/#seq-10.5

**Comment by cyberphone** [comment]
@stefan-kauhaus Since there is no fifth entity (Merchant/Payee) in the depicted schemes, I'm unable to draw any conclusions regarding privacy with respect to the User/Payer.

**Comment by cyberphone** [comment]
@stefan-kauhaus I see.  Personally, I have some difficulties translating the above into a concrete specification.

**Comment by cyberphone** [comment]
A Sequence Diagram with Merchant/Payee
![Sequence Diagram](https://cyberphone.github.io/wallet-core/doc/sequence-diagram.png)

**Comment by tlodderstedt** [comment]
> But wouldn’t it be possible to keep the default OpenID4VP flow and tackle the drawbacks you mentioned in a different way.
> 
> Couldn’t the ASPSP easily detect a replay by tracking the hash of the P2Pay? Once a P2Pay is processed, the ASPSP would refuse the processing if it would be submitted again.

@senexi one time use for replay detection is a rather limited approach as injections and any sort of abuse where the attacker prevents the legitimate party to use the token cannot be prevented. So if we want to make this additional flow a secure option of the normal OpenID4VP flow, we need additional measures in place. 

The difference between normal OpenID4VP and this flow is the fact the RP does not consume the credential presentation itself but rather uses it to authorize an API call with a 3rd party. I suggest to signal this to the wallet in the presentation request through an additional parameter `present_to_thirdparty`, which causes the wallet to respond with an access token wrapping the credential presentation in the `vp_token`.

![image](https://github.com/user-attachments/assets/4c5bccc7-3abf-4808-993a-b34c22cb6c88)

1. (steps 1-4) The TPP (acting as OpenID4VP RP) sends a presentation request with transaction data for payment initiation. In addition to the normal flow, it also indicates that the resulting VC presentation shall be used with a third party (the API of the ASPSP/the bank). It also passes a DPoP key, to be used to proof possession when using the VC presentation (as access token) with the payment initation API of the ASPSP. 
2. (step 8) The wallet asks for consents and pushes the access token (containing the EAA presentation) to the TPP using the TPP's response  endpoint.
3. (steps 9-10) Redirect from wallet back to TPP
4. (step 11) The TPP sends a payment initiation request to the ASPSP (including a DPoP PoP). 
5. ... the rest is payment initiation processing

This is an example request for step 4: 

```json
{
  "client_id": "x509_san_dns:client.psp.org",
  "response_type": "vp_token",
  "redirect_uri": "https://client.psp.org/callback",
  "nonce": "n-0S6_WzA2Mj",
  "dcql_query": {
    "credentials": [
      {
        "id": "A2Pay",
        "format": "dc+sd-jwt",
        "meta": {
          "vct_values": [
            "https://credentials.example.com/a2pay"
          ]
        },
        "claims": [
          {"path": ["id"]},
          {"path": ["payment-product"]},
          {"path": ["payment-uri"]},
          {"path": ["currency"]},
          {"path": ["name"]}
        ]
      }
    ]
  },
  "transaction_data": [
    eyAidHlwZSI6IC...cGluZyBhdCBNZXJjaGFudCBBIiB9
  ],
  "present_to_thirdparty": {
    "credential_ids": [
      "A2Pay"
    ],
    "dpop_key": {
      "jwk": {
        "crv": "P-256",
        "kty": "EC",
        "x": "NASJ2ADuagOvraLf7O4VxcBMbantzL9dd0jpvMLnBfs",
        "y": "OJY6pqCqRIzpEt78OXasWHGgqV5ZGre_3cHtpNH82gg"
      }
    }
  }
}
```

The parameter `present_to_thirdparty` contains the additional data and indicates to the wallet a modified flow. 

* `credential_id` identifies the requested credentials, the RP wants to present to a third party. 
* `dpop_key` is the public key the credential presentation (used as access token) shall be bound to. 

Note: the `present_to_thirdparty` function MAY be combined with transaction data. 

The wallet will ask the user for consent for the transaction and produce the response. The wallet will wrap the respective credential presentation into an JWT-based access token along with a `cnf` for the DPoP key and the `client_id` of the authenticated wallet client. Here is an example access token: 

```json
{
    "iss": "https://issuer.eudiw.com",
    "aud": "https://bank.example.com",
    "client_id": "x509_san_dns:client.psp.org",
    "exp": 1883000000,
    "nbf": 1718198433,
    "iat": 1718198433,
    "credential_presentation": "eyJhbGciOiAiRVMy..V0~Wy..ZnBAan5g",
    "cnf": {
      "jwk": {
        "kty":"EC",
        "x":"l8tFrhx-34tV3hRICRDY9zCkDlpBhF42UQUfWVAWBFs",
        "y":"9VE4jf_Ok_o64zbTTlcuNJajHmt6v9TDVrU0CdvGRDA",
        "crv":"P-256"
      }
    }
}
```

Note: Even though the access token uses the JWT-format, this design is intended to work with _any_ credential format. 

This access token is signed with the same key that is used to sign the credential presentation. The access token is encrypted with a public key of the bank. This key would need to be determined either from data in the A2Pay credential or through metadata the wallet requests from the bank. 

The access token is returned in the `vp_token` instead of the credential presentation in a structure containing the following claims:

* `access_token`: the access token
* `nonce`: the `nonce` as send in the request.
* `metadata`: Additional metadata the wallet shares with the RP. In the case of the A2Pay credential, this is the payment endpoint URI. 

Here is an example:

```json
{
    "vp_token": {
        "A2Pay": {
            "access_token": eyJhbGciOi...00ifQ.OKOaw...zg.48...3b.5eym8T...QD_A.XFBo...SkQ,
            "nonce": "n-0S6_WzA2Mj",
            "metadata": {
                "payment-uri": "https://bank.example.com/pay/7dfe5484g78"
            }
        }
    }
}
```

Note: depending on the use case, the RP might know the API  to send the access token to in advance. In such cases, the RP would add them to the `present_to_thirdparty`structure. If not, the wallet determines those and add it to the response. 

The RP will send this access token along with the DPoP proof in step 11.

**Comment by senexi** [comment]
@tlodderstedt thanks for your proposal. I can see us going forward with that as it is also a more explicit way to handle things. One thing that comes to my mind though is how to determine the `metadata` of the presented credential. One thing would be to just include all disclosed claims also into the meta data property. Another thing would be to flag them somehow during issuing, but this would create a dependency on the wallet itself. What are your thoughts on that?

**Comment by tlodderstedt** [comment]
> What are your thoughts on that?

I have considered to add this to the credential through claims and/or (disclosure) policies. If that is to static, one could also use issuer hosted metadata.

**Comment by tlodderstedt** [comment]
@cyberphone don't you think the PSP could be a merchant? I

**Comment by cyberphone** [comment]
> @cyberphone don't you think the PSP could be a merchant?

Well, Merchants and PSPs represent different roles.  In some cases an entity takes on both roles.  I don't see a need for doing something special protocol-wise for edge cases. I guess that possible optimizations and simplifications can be carried out locally.

**Comment by cyberphone** [comment]
According to https://github.com/nobid-consortium/payment-reference-documentation/blob/main/payment-reference-doc.md#security-considerations Merchants are TPPs that verify.  Verify what?

These guys have another idea on how to use OpenID4VP:
https://ewc-consortium.github.io/eudi-wallet-rfcs/payment-rfcs/implementation-guides/payment-authentication-sca-using-eudi-wallets.pdf#page=38

Facts: 
- Banks already have working PayByBank solutions.  The much hyped https://www.w3.org/TR/secure-payment-confirmation/ tries to compete with already deployed SCA solutions.
- The European Commission is about to launch an Ad-Hoc Task Force for payments using EUDI.
- The majority of the SPRIN-D Funke conference participants thought it is futile competing with Apple and Google.

Any ideas on how to proceed?
@jogu @tlodderstedt @Sakurann

**Comment by mobinauten** [comment]
@cyberphone  We just sync. with EWC 48hs ago and they are very much aligned to this proposal. We both sync regulary.

**Comment by mobinauten** [comment]
@cyberphone Merchandts verify age e.g., shipping adresses @cyberphone  Why don't you join one of the parties in NOBID or EWC to contribute directly and get a very deep insight into the current software developments? 
e.G in NOBID one of the main actors in REWE, a large retailer in Germany,  having implemented the first versions of these approaches, so an experienced merchant is on board and provides hands-on experiences and developers.

**Comment by mobinauten** [comment]
And the the LSPs are part of the ATAG.

**Comment by mobinauten** [comment]
And it was from my perspective the wrong question at the SPRIND conference. This approach is not to compete against Apple, Google, Mastercard or Visa. 
It should enable payment in an EUDIW in an open, standardized and scheme-agnostic way in order to combine payment and credentials/data in one app/transaction and to provide superb CX. 
It's not about inventing the next scheme. 
And maybe Google or Apple will develop an EUDIW as well, who knows? Or Visa? Then this approach should work as well. 
ieIDAS2 is part of a bigger EU infrastructure picture and thus we don't need to ask the competition question, I believe.

Ideas how to proceed. Very clear. 

1. Continue work on LSP1 
2. Start work on LSP2s 
3. Synchronize with national developments like FUNKE in DE
4. Synchronize with the EU and national and EU banking authorities

**Comment by senexi** [comment]
@cyberphone 
> According to https://github.com/nobid-consortium/payment-reference-documentation/blob/main/payment-reference-doc.md#security-considerations Merchants are TPPs that verify. Verify what?

As stated in the document: "verification of the P2Pay". In case you are asking what a P2Pay is... it is explained in the document.

> These guys have another idea on how to use OpenID4VP: https://ewc-consortium.github.io/eudi-wallet-rfcs/payment-rfcs/implementation-guides/payment-authentication-sca-using-eudi-wallets.pdf#page=38

Seriously? Do you actually read these documents? Maybe you would notice that they are quite similar....

https://github.com/EWC-consortium/eudi-wallet-rfcs/blob/main/payment-rfcs/ewc-rfc008-payment-data-confirmation.md

But I would like to focus on the actual issue and proposal from @tlodderstedt here...

Regarding the metadata, I think it absolutely makes sense to move some of the claims from the credential to the PSP / Issuers metadata, as they are the same for every issued credential. Making the approach more flexible and reducing the static data within the credential. @stefan-kauhaus also pointed out that option.

**Comment by cyberphone** [comment]
@mobinauten 
> And it was from my perspective the wrong question at the SPRIND conference. This approach is not to compete against Apple, Google, Mastercard or Visa.

Personally, I don't see much point in developing a system that doesn't have these characteristics.  The system I'm working with (entirely on my own and without any funding 😪), is explicitly designed for competitiveness.

> Merchandts verify age e.g., shipping adresses

Yes, but this is unrelated to payments.

> Why don't you join one of the parties in NOBID or EWC to contribute directly and get a very deep insight into the current software developments?

I believe I did but I got no response.  Anyway, without clearly separating Merchants and TPPs, I wouldn't consider joining the party.

**Comment by c2bo** [comment]
I think being explicit that this flow (whatever it will look like) is intended to be shown to / used with a third party would be a very good idea here -> +1 on the `present_to_thirdparty` proposal or something similar.

**Comment by cyberphone** [comment]
@c2bo This is implicit in other payment authorization systems.  Maybe the idea is that a single wallet implementation, should be able to handle everything from age-assertions to payments? This would IMO be a pretty bad idea.

My guess is that Apple have a common code-base for certain functions like crypto, but that the wallet dispatches requests to application-specific "executors".

Anyway, a unified flow seems pretty distant since we (apparently) have quite different ideas about what a Merchant is.  In the depicted schemes, Merchants would either need to have contracts with all issuer banks or be accredited by an NCA (National Competent Agency) as an PSD2 entity.  This obviously doesn't apply to the coffee shop on the corner. There is a reason to why companies like Adyen, Stripe, Worldline exists.  They handle millions of Merchants and thousands of Banks.

There ARE ways to change this but this require a major architectural update.  I'm not actively pushing this but it might be interesting to see anyway...

![saturnflows li](https://github.com/user-attachments/assets/be58bd62-f950-4a59-aac5-ce990cf96478)

**Comment by cyberphone** [comment]
@tlodderstedt @senexi @mobinauten The bigger question (IMO...) is if we believe we are creating a standard for an EUDI Wallet with payment functionality.  I don't think we are even close because we haven't yet concluded what our expectations are.  Such a standard would also have to (in order to be "authoritative") be developed by a standards committee.  OpenID does currently not have a "Payment Wallet WG".

Is this a problem?  Yes, the lack of a real standard will undoubtedly stall large-scale deployments unless one of the initiatives will considered a clear winner.

Personally (FWIW), I have limited faith in Berlin Group's take on how to support payment wallets which caused me to develop an entirely different solution.  Currently it is just "slideware", but I actually did a similar solution with a real bank back in 2019, and it worked pretty well.  The main points with this concept are:
- Reducing the need for centralized standardization of gazillions of APIs.
- Permit third-parties creating application-specific APIs that (at least in theory), with moderate efforts would work in any number of banks.
- Enabling competition by wallet designs.

https://github.com/cyberphone/open-banking-2.0

![OB2](https://cyberphone.github.io/open-banking-2.0/doc/ob2-overview.png?)

**Comment by mobinauten** [comment]
Indeed, we should differentiate between the enabling infrastructure and the payment mean and architecture as well. Different layers. 
The approach in NOBID and EWC and thus in OIDF could be/is already to creating the necessary rails/layers that then can be used by the different payment architectures. And these payment means and architectures then very likely will use other standards as well like from W3C or ISO or...
The EUDIW and OIDF are not, should not be Payment means with all their rulebooks like cards or A2A schemes. 
It could powers these but are not.

**Comment by cyberphone** [comment]
Maybe we could take a step back and define a place for the Merchant?

In my take on payment authorizations (which predates OpenID4VP with 8 years or so), the wallet is entirely unaware of rails and specific payment networks, it is just a fairly simple Merchant payment request-response scheme.  There is no credential presentation or verification either; the wallet response only provides core data needed for payment network and bank selection.

It is not clear to me what more specifically an SDO would do.  Define the content of `transaction_data`?  I hope not.

Current specifications also need to be tested before any meaningful standardization can be performed.  We are probably talking 2030 before such a standard would be ready (and probably already obsolete).

VIP: https://cyberphone.github.io/wallet-core/doc/

**Comment by mobinauten** [comment]
@cyberphone We're testing these approaches since more than 12 months in LSPs NOBID and EWC. You should really join one of those a the coming ones to get deeper insights. 
Cheers Olli 
P.S SDO is what?

**Comment by cyberphone** [comment]
> We're testing these approaches since more than 12 months in LSPs NOBID and EWC.

Dear @mobinauten No one doubts that your stuff works.  My questions are related to scalability and architecture.

> You should really join one of those a the coming ones to get deeper insights.

At this stage I prefer sticking to the publicly available documents.  I'm sort of a spec. nerd. 😆. EWC seems to also target 3D Secure, something I would never consider since it is already available, albeit (typically) through through mobile banking applications.

SDO is short for Standards Development Organization.

FWIW, I have limited faith in Berlin group's Signed Payment Request.  Their single-level "monster" API concept, makes development and testing a nightmare!

Cheers Anders

**Comment by cyberphone** [comment]
@mobinauten When it comes to data which is entirely _Merchant-related_ like _Age verification_, _Loyalty_, and _Physical address_, I believe it would be a problem trying to squeeze this into a payment request message.  AFAICT, these items could (without any UX disadvantages for the user), be requested _before_ the payment request and presumably using the EUDIW "as is".  Well, Loyalty is something I don't think has been covered.  Although loyalty could be obtained though identity (authentication) this may not be an ideal solution since the identity is targeted for a single Merchant/Brand.  Giving your IKEA card to LIDL should not be possible!  The WebAuthn security model (domain separation) seems pretty appropriate for Loyalty.  Maybe this is already a part of EIDIW?

**Comment by cyberphone** [comment]
Regarding NOBID/EWC alignment, EWC seems to also have support recurring payments: https://github.com/EWC-consortium/eudi-wallet-rfcs/blob/main/payment-rfcs/ewc-rfc008-payment-data-confirmation.md#41-authorisation-request

Although cool, does the EWC solution actually support gas station payments?

**Comment by mobinauten** [comment]
@cyberphone  What issues. Could you maybe be more precise and reference examples in the context of the current proposals. This would help. Thanks Olli

**Comment by cyberphone** [comment]
@mobinauten Well, there are virtually tons of issues.  The sequence diagrams above do not include Merchants.  Maybe the whole payment authorization flow is supposed to be outsourced?  I'm just guessing here.  Even if payment authorization is outsourced, doesn't it still require some kind of trusted link be between the Merchant and the payment intermediary?

Somewhat related: the UK (OBL, FCA, CMA) claim that VRP (Variable Recurring Payments) is all that is required to make Open Banking payments flourish.  I certainly do not agree with that; VRP is a _part_ of the solution, not _THE_ solution.

**Comment by mobinauten** [comment]
@cyberphone ? The TPP/Verifier is the merchant above? Maybe you also look at the sequence diagrams @tlodderstedt has created.

**Comment by cyberphone** [comment]
@mobinauten @tlodderstedt  To begin with there are some terminology issues making the specifications hard to follow.  For payment folks, "OAuth client" does not sound like a "Merchant".  In Open Banking contexts using PSD2 terms, an "OAuth client" is usually a PISP and/or AISP.

Similar issues exist with the OpenID4VP "Verifier" concept because there are many different verifications involved in a payment transaction like:
- Verification of the User's authorization
- Verification of the Merchant's authenticity
- Verification of the Bank's response to the Merchant

I totally understand if you think the following is overkill.  However, since none of us can [rightfully] claim to know everything, I believe it could be useful identifying the actors and their roles and trust relationships.  The Merchant role seems a bit underspecified.  Merchants are not relying parties, at least not in the traditional sense.  In my world, Merchants want your money and that's all they should get as well 😎. Sure, there are exceptions but personally I believe this should be kept outside of the payment process.

**Comment by cyberphone** [comment]
@tlodderstedt @jogu Narrowing down things a bit...

According to you, is a Merchant that is also a TPP (=performs payment initiation)
- A new kind of entity?
- An entity which takes on two distinct roles?

You may wonder what this has to do with signed `transaction_data`, right?

Well, with _distinct_ roles there is no apparent need for signing the request to the user; the TPP deals with Merchant authentication as well as with `creditorAccount` data.  Through a `hostName` object gathered by the wallet during invocation, a TPP can securely bind user authorization responses to a Merchant.  This is effectively a copy of the FIDO/WebAuthn security concept.  Lightweight, scalable, and simple.

**Comment by cyberphone** [comment]
There are other things in https://github.com/openid/OpenID4VP/issues/342#issuecomment-2547758170 that are confusing.  "User Agent" usually means browser.  From the sequence diagram it seems that the browser displays the approval dialog.  This is probably not what the EUIDW folks had in mind. Is the EUIDW in the sequence diagram a server of some kind?

**Comment by Sakurann** [comment]
the feedback seems to have been that the need for feature is gone

**Comment by senexi** [comment]
Closing this issue for now, as the latest assessment has shown that we can move forward building on the client_id which is included in the KB-JWT of the presentation.

**Comment by cyberphone** [comment]
@senexi @mobinauten @tlodderstedt @jogu @Sakurann @stefan-kauhaus 
I must confess that I don't understand much of this.  Creating a complex multi-party information system without first identifying
- the entities
- their roles
- their trust relationships
- the information model
- the privacy model
- the security model

seems like a lost case.

BTW, why would payment authorization use selective disclosure?  This must be the result of a confusion regarding the (non-existing) information model, or that you are trying to combine operations that other systems implement as separate processes.  Loyalty information as an example MUST proceed payment requests since it might affect the price.  Apple have recently introduced transparent loyalty card support.  They certainly do not send this information to the bank.

Anyway, these issues are insignificant compared to https://github.com/openid/OpenID4VP/issues/429

**Comment by cyberphone** [comment]
Apparently the concept has changed dramatically:
https://github.com/digitallabor-berlin/eudiw-sca/blob/ea5db12b59f583f79e3c866896565fa6c93ae2e4/openbanking-r2s.md#payment

<img width="1000" alt="Image" src="https://github.com/user-attachments/assets/547c7e3f-847f-4a10-98fd-51f9001c3871" />

This is a better solution than the one shown above since it doesn't require Merchants to be certified by an external body.

A more developed version of this concept can be found here: https://github.com/eu-digital-identity-wallet/eudi-doc-architecture-and-reference-framework/issues/388
